zerver/lib/markdown/fenced_code.py

```diff
@@ -252,16 +253,29 @@ def __init__(
         self.lang = lang
         self.lines: List[str] = []
         self.run_content_validators = run_content_validators
+        self.need_to_handle_unclosed_fence = False
 
-    def handle_line(self, line: str) -> None:
+    def handle_line(self, line: str) -> bool:
+        rstrip_line = line.rstrip()
+        line_length = len(rstrip_line)
+        line_has_only_fence = line_length >= 3 and (
+            rstrip_line.count("`") == line_length or rstrip_line.count("~") == line_length
+        )
         if line.rstrip() == self.fence:
             self.done()
+            self.need_to_handle_unclosed_fence = False
+            return self.need_to_handle_unclosed_fence
+        elif not self.need_to_handle_unclosed_fence and line_has_only_fence:
+            self.need_to_handle_unclosed_fence = True
+            return self.need_to_handle_unclosed_fence
         else:
             self.lines.append(line.rstrip())
+            self.need_to_handle_unclosed_fence = False
+            return self.need_to_handle_unclosed_fence
```

Comment:
We definitely don't want to copy this block of logic into 4 different classes -- we should do whatever preparatory refactor commit lets us avoid doing so.

Reply:
Okay, I will refactor this. Right now, I just submitted this to show what I'm thinking of a possible solution. I will clean it in future PR.